### ISSUE TITLE
Improve Caddy path matching

### ISSUE THREAD

**Issue by dhaavi** [mholt/caddy#1862]
I am currently using Caddy on multiple production server in different settings. So far I really love the simplicity, speed and stability of Caddy. In the last weeks I have also experienced two limits of Caddy. Both of them are, I think, very important to discuss, as they may become a dealbreaker easily. In order to secure the future Caddy (you want companies to buy [these plans](https://caddyserver.com/pricing), right?), these issues should be thoroughly discussed - as they are also breaking changes. 

The first issue is about path matching and has also been partly discussed in #167. If there are other similar issues, or this is already currently being discussed in another issue (I couldn't find anything current), please tell me.

This is not beautiful:

    www.example.com {
      proxy /en 10.0.0.1:80
      redir 301 {
        if {path} not_starts_with /en
        / www.example.org
      }
    }

This is:

    www.example.com {
      proxy /en 10.0.0.1:80
      redir / www.example.org
    }

This is just a small example, the underlying issue is much more important and complex. How I experienced Caddy, plugins are asked sequentially if they handle a certain path, if they do, the path matching phase ends, no matter how many plugins would have been left to check. Correct me if I am wrong - if I am, please explain what creates this behaviour.

Looking at the example, I had to exclude the subsite proxy directive from redirection, as apparently `redir` is processed before the `proxy` plugin. In this case there was a solution, but with lots of plugins, you will quickly encounter situations where a plugin will not have the ability to exclude a path. Also, the process of finding the correct configuration is a lot of effort and the resulting configuration is ugly.

This behavior makes writing Caddy config needlessly complex, as I have to know in which order plugins are being called and have to exclude all paths from earlier plugins that I want later plugins to handle.

I would suggest to handle paths similar to nginx does with its [location](https://nginx.org/en/docs/http/ngx_http_core_module.html#location) directive, as it works really well and I guess A LOT of Caddy users will already know nginx and how paths work there, especially if you are trying to attract companies to use Caddy. (Just a quick side thought: a nginx to Caddy config converter - at least for simple things - may help in getting more users)

Though the most important thing is that the Caddy core handles path matching and not the plugins directly.

Until this issue is resolved, please update to documentation to explain how the current path matching works exactly. While searching for a solution for my use case I did not find any piece of documentation that was helpful in this sense and discovered the solution by chance. This is not what Caddy stands for.

Whatever you do, please do one thing: keep up this awesomeness.

**Comment by mholt** [comment]
Hi, thanks for your comments!

First, I want to clarify something:

> In order to secure the future Caddy (you want companies to buy these plans, right?) ...

Companies can buy the Engineering Package if they want their feature requests like this prioritized, yes. :wink: At this point, though, I do not rely on the future of Caddy. I hope it will be successful and do good in the whole world to improve the Web for both developers and every user of the Internet, but if Caddy stops being maintained, I'll just find other projects to work on or focus on a full-time job instead. It's all the people relying on Caddy that I would think would have a rough transition. Which is why next week we're changing our product offering. The imbalance of Caddy's growth & popularity vs. the amount of time I can afford to put into it is not healthy for the ecosystem, so next week I'm hoping we'll have options that make the whole thing better for everyone.

Okay, back to the main point:

I agree Caddy path matching needs to improve, but probably not until after version 1.0. And we won't limit it to matching paths: we'll match _requests_ which can look at more than just the path. I hope to have a standard way of doing this that all middleware can use, because right now, as you said, different middleware match requests differently, and it can be a little confusing. I wish I had standardized this better right from the get-go.

However, this is a duplicate of #619 which has the very first point is "More powerful path matching" -- but again, it'll actually be _request matching_.

> This behavior makes writing Caddy config needlessly complex, as I have to know in which order plugins are being called and have to exclude all paths from earlier plugins that I want later plugins to handle.

No matter how we/they do it, there will have to be some order in how things are evaluated. As I understand it, nginx has to evaluate location blocks in some order and then it even re-processes requests until rewrites stop. In the end, the code needs to execute in _some_ order. With Caddy, you don't really have to think about the order. To be fair, what you want as far as syntax:

```
www.example.com {
  proxy /en 10.0.0.1:80
  redir / www.example.org
}
```

is no less confusing and ambiguous. There's no reason that this is the "right" way to specify your configuration. It's just how Caddy is implemented. In other words, who says that proxy should be exclusive of redir? It's just how it's implemented. You want more specific paths to have priority, but what if we want to match requests on a combination of remote IP, header field values, or time of day?

People tend to pressure Caddy to be more like nginx or apache because that's what they're used to, but I'm not inclined to go that route. Caddy will continue to innovate in ways that help push the web forward, rather than locking us into one line of thinking.

**Comment by mholt** [comment]
Also, your config example is a bit of a straw man argument, since this is the "Caddy way" (that you refer to):

```
www.example.com {
    redir / www.example.org
}

www.example.com/en {
    proxy / 10.0.0.1:80
}
```

Which is much nicer IMO.

**Comment by dhaavi** [comment]
Thank you for your reply. I am then looking forward to next week! :)
Request matching sounds fun!

Oh, hey! I did not realize that was possible, thanks for the hint.

**Comment by mholt** [comment]
Request matching is hard. It'll require a complete overhaul of Caddy's fundamental design, maybe its architecture. The Caddyfile will probably change too. Open to ideas on that...

**Comment by dhaavi** [comment]
I'd actually love to dive into the Caddy code and think about this more, but sadly I currently have so much to do that this won't be possible for at least half a year. I will sure stick around though ;)

**Comment by mholt** [comment]
Okay, no problem. Thanks for your input! (And to clarify, a closed issue doesn't mean discussion isn't allowed -- it's just because there's nothing specific on the TODO list from it.)

**Comment by dhaavi** [comment]
Sure - I understand ;)
